### PR TITLE
Nemesis configuration improvements

### PR DESCRIPTION
**Pull request by fboucquez**
~1. Fixed config/start --preset and --assembly options (it was the other way around)~ Already merged in another pr
2. Removed unnecessary currency names from custom preset. It can be provided and resolved from the `nemesis.mosaics` field
3. Moved "opt-in" addresses/balances from `nemesis` to `nemesis.mosiac`. User (or symbol-network) can provide balances for multiple mosaics, although most of the time there is only one nemesis mosaic (like xym).
4. Allow a node to be excluded from nemesis (no balance or linked transactions). 
5. Allow users to provide node nemesis balance. If the balance is calculated, it will be less than  `maxHarvesterBalance` 
6. Better nemesis supply distribution and exception handling. Added unit tests.  

If this is too much I can split further. Note I would be creating a new PR where the genesis related configuration is moved to their own class/service

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -186,6 +185,12 @@ Although some combinations can be done, they may not be really useful. Examples
 -   `$ symbol-bootstrap start -p mainnet -a multinode`
 -   `$ symbol-bootstrap start -p testnet -a multinode`
 
+A custom network preset file can also be provided. This is useful when you have your own custom Symbol network, and you want other nodes to join.
+For this case, you provide your own `networkPreset.yml` and nemesis feed folder. The node admin can then run:
```
Comment (segfaultxavi) [inline README.md:189]:
```suggestion
For this case, you provide your own `networkPreset.yml` and nemesis seed folder. The node admin can then run:
```

File: src/commands/config.ts [hunk 2]
```diff
@@ -33,15 +35,17 @@ export default class Config extends Command {
         target: CommandUtils.targetFlag,
         password: CommandUtils.passwordFlag,
         noPassword: CommandUtils.noPasswordFlag,
-        preset: flags.enum({
+        preset: flags.string({
             char: 'p',
-            description: `The network preset. It can be provided via custom preset or cli parameter. If not provided, the value is resolved from the target/preset.yml file.`,
-            options: Object.keys(Preset).map((v) => v as Preset),
+            description: `The network preset. It can be provided via custom preset or cli parameter. If not provided, the value is resolved from the target/preset.yml file. Options are: ${Object.keys(
+                Assembly,
```
Comment (segfaultxavi) [inline src/commands/config.ts:41]:
```suggestion
                Preset,
```

File: src/commands/config.ts [hunk 3]
```diff
@@ -33,15 +35,17 @@ export default class Config extends Command {
...
+            ).join(', ')}, my-custom-network.yml (advance)`,
         }),
-        assembly: flags.enum({
+        assembly: flags.string({
             char: 'a',
-            description: `The assembly that defines the node(s) layout. It can be provided via custom preset or cli parameter. If not provided, the value is resolved from the target/preset.yml file.`,
-            options: Object.keys(Assembly).map((v) => v as Assembly),
+            description: `The assembly that defines the node(s) layout. It can be provided via custom preset or cli parameter. If not provided, the value is resolved from the target/preset.yml file. Options are: ${Object.keys(
+                Preset,
```
Comment (segfaultxavi) [inline src/commands/config.ts:47]:
```suggestion
                Assembly,
```

File: src/commands/config.ts [hunk 4]
```diff
@@ -33,15 +35,17 @@ export default class Config extends Command {
...
         password: CommandUtils.passwordFlag,
         noPassword: CommandUtils.noPasswordFlag,
-        preset: flags.enum({
+        preset: flags.string({
             char: 'p',
-            description: `The network preset. It can be provided via custom preset or cli parameter. If not provided, the value is resolved from the target/preset.yml file.`,
-            options: Object.keys(Preset).map((v) => v as Preset),
+            description: `The network preset. It can be provided via custom preset or cli parameter. If not provided, the value is resolved from the target/preset.yml file. Options are: ${Object.keys(
+                Assembly,
+            ).join(', ')}, my-custom-network.yml (advance)`,
```
Comment (segfaultxavi) [inline src/commands/config.ts:42]: What does `(advance)` mean?

**Review by segfaultxavi — CHANGES_REQUESTED** [review]
Some mixed-up options...

File: src/commands/config.ts (hunk 2 above)
Comment (fboucquez) [reply src/commands/config.ts]: Thanks!

File: src/model/ConfigPreset.ts [hunk 5]
```diff
@@ -259,6 +264,7 @@ export interface NodePreset extends DockerServicePreset, Partial<NodeConfigPrese
     host?: string;
     roles?: string;
     friendlyName?: string;
+    excludeFromNemesis: boolean;
```
Comment (Jaguar0625) [inline src/model/ConfigPreset.ts:267]: why is this option needed?

File: src/service/ConfigLoader.ts [hunk 6]
```diff
@@ -130,43 +130,107 @@ export class ConfigLoader {
...
+                return d.amount;
+            })
+            .reduce((a, b) => a + b, 0);
+    }
+    private processNemesisBalances(presetData: ConfigPreset, addresses: Addresses, nemesisSignerAddress: Address): MosaicAccounts[] {
+        const privateKeySecurityMode = CryptoUtils.getPrivateKeySecurityMode(presetData.privateKeySecurityMode);
+        const networkType = presetData.networkType;
+        const mosaics: MosaicAccounts[] = [];
+        presetData.nemesis.mosaics.forEach((m, mosaicIndex) => {
+            const accounts = this.generateAddresses(networkType, privateKeySecurityMode, m.accounts || 1);
```
Comment (Jaguar0625) [inline src/service/ConfigLoader.ts:159]: why || 1 ?

File: src/model/ConfigPreset.ts (hunk 5 above)
Comment (fboucquez) [reply src/model/ConfigPreset.ts:267]: One use case,

When symbol-networks creates the 'genesis block' node (creating the network nemesis block), all the balances and link transactions are explicitly provided by the tool (according to the network layout defined in a previous step). I don't want the nemesis block node to be included at all in the nemesis block.

Comment (fboucquez) [reply src/model/ConfigPreset.ts:267]: This step: https://hackmd.io/Ak7Sl1A6SkKV7ga0QW00LA#Nemesis-generation

File: src/service/ConfigLoader.ts (hunk 6 above)
Comment (fboucquez) [reply src/service/ConfigLoader.ts]: Removed and unit tested

File: src/model/ConfigPreset.ts (hunk 5 above)
Comment (Jaguar0625) [reply src/model/ConfigPreset.ts:267]: what is this use case? where did this requirement come from?

Comment (fboucquez) [reply src/model/ConfigPreset.ts:267]: It's a helper ffeature for my symbol-network tool

Comment (Wayonb) [reply src/model/ConfigPreset.ts:267]: I see the ``excludeFromNemesis`` property getting use [here.](https://github.com/symbol/symbol-testnet/blob/deb438f6bf5114136af6b89b573c59f2f70d6e33/joeynet/nemesis-target/preset.yml#L775) 
Is this just to create local nodes?

Comment (fboucquez) [reply src/model/ConfigPreset.ts:267]: Yes, it's used by the tool in one of the steps. The step is when the local "nemesis" node is created. That nemesis node generates the nemesis block where all the other final node balances and transactions are provided. I don't want to include anything to that node in the nemesis block so the exclude flag.

Comment (Jaguar0625) [reply src/model/ConfigPreset.ts:267]: are you saying that there's a local node that is generating all the other nodes?

Comment (fboucquez) [reply src/model/ConfigPreset.ts:267]: The local node's nemesis block (seed folder) is copied over the other nodes when they are generated.

**Review by Jaguar0625 — APPROVED** [review]

**Pull request merged**